parachain/primitives/router/src/outbound/mod.rs

```diff
@@ -1492,4 +1591,108 @@ mod tests {
 		let result = converter.convert();
 		assert_eq!(result.err(), Some(XcmConverterError::BeneficiaryResolutionFailed));
 	}
+
+	#[test]
+	fn is_transact() {
+		let network = BridgedNetwork::get();
+
+		let fee = MultiAsset { id: Concrete(Here.into()), fun: Fungible(1000) };
+		let fees: MultiAssets = vec![fee.clone()].into();
+
+		let message: Xcm<()> = vec![
+			WithdrawAsset(fees),
+			BuyExecution { fees: fee.clone(), weight_limit: Unlimited },
+			Transact {
+				origin_kind: OriginKind::Native,
+				require_weight_at_most: Default::default(),
+				call: vec![].into(),
+			},
+			SetTopic([0; 32]),
+		]
+		.into();
+
+		let converter = XcmConverter::new(&message, &network, &GATEWAY);
+
+		let result = converter.is_transact();
+		assert_eq!(result, true);
+	}
+
+	#[test]
+	fn is_not_transact() {
+		let network = BridgedNetwork::get();
+
+		let fee = MultiAsset { id: Concrete(Here.into()), fun: Fungible(1000) };
+		let fees: MultiAssets = vec![fee.clone()].into();
+
+		let token_address: [u8; 20] = hex!("1000000000000000000000000000000000000000");
+		let beneficiary_address: [u8; 20] = hex!("2000000000000000000000000000000000000000");
+
+		let assets: MultiAssets = vec![MultiAsset {
+			id: Concrete(
+				X2(
+					AccountKey20 { network: None, key: GATEWAY },
+					AccountKey20 { network: None, key: token_address },
+				)
+				.into(),
+			),
+			fun: Fungible(1000),
+		}]
+		.into();
+
+		let filter: MultiAssetFilter = Wild(WildMultiAsset::AllCounted(1));
+
+		let message: Xcm<()> = vec![
+			WithdrawAsset(fees),
+			BuyExecution { fees: fee.clone(), weight_limit: Unlimited },
+			WithdrawAsset(assets),
+			DepositAsset {
+				assets: filter,
+				beneficiary: X1(AccountKey20 {
+					network: Some(Ethereum { chain_id: 2 }),
+					key: beneficiary_address,
+				})
+				.into(),
+			},
+			SetTopic([0; 32]),
+		]
+		.into();
+
+		let converter = XcmConverter::new(&message, &network, &GATEWAY);
+
+		let result = converter.is_transact();
+		assert_eq!(result, false);
+	}
+
+	#[test]
+	fn transact_message_unpaid_execution() {
+		let network = BridgedNetwork::get();
+
+		let fee = MultiAsset { id: Concrete(Here.into()), fun: Fungible(1000) };
+		let fees: MultiAssets = vec![fee.clone()].into();
+
+		let message: Xcm<()> = vec![
+			UnpaidExecution { weight_limit: Unlimited, check_origin: None },
+			DescendOrigin(X1(AccountKey20 {
+				network: Some(NetworkId::Ethereum{ chain_id: 15 }),
+				key: hex!("1000000000000000000000000000000000000000"),
+			})),
```

Comment:
@yrong  I need to send the Ethereum contract address as part of the XCM message. It does not look like it would form part of the Transact XCM message. `DescendOrigin` looked like the closest thing that matched, let me know if this is OK.

Reply:
Yes, you can do that. But I think the `Gateway` address is already hardcode as a const into the runtime [here](https://github.com/Snowfork/snowbridge/blob/30be20dca6a505fe7c3c02f1d496b60c967f4cb5/parachain/primitives/router/src/outbound/mod.rs#L31) so you may not need it.

Reply:
Ah, it's the destination contract not the gateway. Yes, definitely need it.